### ISSUE TITLE
$wpdb->insert() is not a direct replacement for $gpdb->insert()

### ISSUE THREAD

**Issue by toolstack** [GlotPress/GlotPress#52]
$gpdb->insert() does some additional processing that is not done in $wpdb->insert().

Particularly troublesome is the fact that it substitutes NULL for data values that are not set, which GlotPress then checks for in various places in SELECT statements.

For example, start a new project and import some originals, if no context or plural are set, each original will have 'context' and 'plural' set to an empty string not NULL.

Then create a translation set and import some strings, no values are imported as the import code try's to find the originals with NULL values in the 'context' and 'plural' fields.

There are a couple of solutions off the top of my head:
1. Continue to use $gpdb but instead of using the backpress version create our own that is an extension of $wpdb.
2. Find all the select statements that check for null and replace/add an empty string to them as well.

Any one else have any thoughts on this?

**Comment by ocean90** [comment]
Good point! I think this will be fixed in 4.4, see https://core.trac.wordpress.org/changeset/34737.

**Comment by toolstack** [comment]
I'll give the trunk $wpdb class a try and see if it resolves it, if not I've been looking through the code and there aren't too many places where it would be impacted to add in a check for NULL and an empty string.

**Comment by toolstack** [comment]
Yep, trunk's version works fine, I guess GlotPress as a Plugin is going to require WP 4.4 or above :smile:

**Comment by ocean90** [comment]
>  I guess GlotPress as a Plugin is going to require WP 4.4 or above 

Yep, and I think that's OK.

**Comment by toolstack** [comment]
I agree, 4.4 should be fine, I don't see any reason to do a lot of work getting it to function pre-4.4 when 4.4 is so close anyway.

**Comment by toolstack** [comment]
Here are some additional thoughts on the warning message PR, there are two areas we should think about:
1. Installing V1.0.0 on existing WordPress installs.
2. Installing/Upgrading VFuture on various versions of WordPress.

These items can be broken down even further:
1. Installing V1.0.0 on existing WordPress installs.  
   1.1 Installing V1 on WP < 4.4  
   1.2 Installing V1 on WP >= 4.4  
2. Installing/Upgrading VFuture on various versions of WordPress.  
   2.1 Installing VFuture on WP < Supported  
   2.2 Installing VFuture on WP >= Supported  
   2.3 Upgrading to VFuture on WP < Supported  
   2.4 Upgrading to VFuture on WP >= Supported  

We can safely ignore 1.2, 2.1 and 2.4, which leaves us with:

### Installing V1 on WP < 4.4 & Installing VFuture on WP < Supported

When a user installs GP on their WordPress install for the first time they should pay attention to the install requirements, but may not for a variety of reasons:
- Not everyone pays attention to the "Requires" on wordpress.org's plugin directory
- May not get the source from wordpress.org but instead from somewhere else
- Manual download from wordpress.org and upload to their site
- May want to try it anyway as lots of plugins say "requires" but actually mean "not supported but might work"

In these cases, GP would install successfully ("Requires" is not enforced by the installer) without any errors.  The majority of the plugin will work, it would not be until they started trying to import originals, add translations and other functions that things would start to go wrong.

In this case I think it is better to gracefully fail with a warning message, providing information about what's wrong, to the user than let them get in too deep and then find out things are broken.

This also ensures we don't have users mad at us for providing a "broken" plugin, even though it isn't.

### Upgrading to VFuture on WP < Supported

The second scenario we will likely run in to at some point in the future (probably far future but we should think about it) is when we need to bump the minimum WP version we require.

In this case users will get prompted to update GP (or even worse, have enabled auto plugin updates) without a notice (or very minimal notice, the upgrade notes do get displayed but my experience so far is that nobody reads them) that are not running a supported version of WordPress. 

In the worst case scenario there could have been a breaking change that we can't support both the new and old versions of WP at the same time, which might result in corruption of data.  It would be better to "gracefully" fail at this point.

This is of course a reasonable unlikely scenario, but a good thing to think about now instead of later.

### What about the PHP version

Of course much of this also applies to the fact we require PHP 5.3 and WP supports 5.2.  I've run in to this with one of my plugins that requires PHP 5.3.  

I added additional code to check the PHP version and made sure to alert the user to it.  This avoids many (not all unfortunately) questions about why the plugin doesn't work.

It also avoid white screens as the PHP 5.3 code would otherwise break the WordPress page load.

### To sum up

GP Standalone can set it's requirements quite specifically and since it has a high barrier to entry, it won't run in to many problems with people installing on unsupported software stacks.

GP as a plugin on the other hand is going to be in the wordpress.org plugin directory (presumably at least :wink:) and be a single click install for anyone who might want to give it a try.

Playing it safe and throwing a warning making sense to me.
